sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -619,3 +620,43 @@ object DataSourceStrategy {
     (nonconvertiblePredicates ++ unhandledPredicates, pushedFilters, handledFilters)
   }
 }
+
+
+/**
+ * Support for recalculating table statistics if table statistics are not available.
+ */
+class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
+  def getHdfsSize(catalogTable: CatalogTable): Long = {
+    try {
+      val hadoopConf = session.sessionState.newHadoopConf()
+      val tablePath = new Path(catalogTable.location)
+      val fs: FileSystem = tablePath.getFileSystem(hadoopConf)
+      fs.getContentSummary(tablePath).getLength
+    } catch {
+      case e: IOException =>
+        logWarning("Failed to get table size from hdfs.", e)
+        session.sessionState.conf.defaultSizeInBytes
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case logical @ LogicalRelation(relation, _, Some(table), _)
+      if session.sessionState.conf.fallBackToHdfsForStatsEnabled &&
+        session.sessionState.conf.manageFilesourcePartitions && table.stats.isEmpty &&
+        table.tracksPartitionsInCatalog && table.partitionColumnNames.nonEmpty =>
+      val withStats =
```

Comment:
Recalculated only when `fallBackToHdfsForStatsEnabled` and `useCatalogFileIndex`:
https://github.com/apache/spark/blob/c30b5297bc607ae33cc2fcf624b127942154e559/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L379-L388